### PR TITLE
Bug 1866605: metadata.generation not getting set/updated

### PR DESCRIPTION
**Pull request by coreydaley-redhat**
Updating to set metadata.generation on create, and increment metadata.generation on update
Added tests for create/update metadata.generation

### CODE REVIEW

**Review by gabemontero — COMMENTED** [review]
Interestingly, `Generation` is also dealt with for DCs, imagestreams, templates in their respective controllers in OCM.

I do see similar setting for those types in openshift-apiserver like what @coreydaley-redhat is doing here

But I'm curious if we need to consider how Generation is touched upon in OCM for those other types and see if any of those patterns apply for builds and buildconfigs as well.

Or @coreydaley-redhat @adambkaplan did you all already go through that exercise while I was out?

**Comment by coreydaley-redhat** [comment]
@gabemontero I based this off of what was being done in the DeploymentConfigs

**Comment by gabemontero** [comment]
> @gabemontero I based this off of what was being done in the DeploymentConfigs

Right I saw that.

But along those lines, I wonder if we should mimic for BCs etc. what is done with Generation for DeploymentConfigs in the OCM as part of completing this bug.

**Comment by coreydaley-redhat** [comment]
@gabemontero I suppose we could, right now the only thing that seems to get updated in the OCM is the LastVersion for the BuildConfigs.  Maybe if Generation is not currently used for Builds and BuildConfigs it is a metric that we should not be trying to report?!?

File: pkg/build/apiserver/registry/buildconfig/strategy.go [hunk 1]
```diff
@@ -70,6 +73,10 @@ func (strategy) PrepareForUpdate(ctx context.Context, obj, old runtime.Object) {
 	if newBC.Status.LastVersion < oldBC.Status.LastVersion {
 		newBC.Status.LastVersion = oldBC.Status.LastVersion
 	}
+
+	if !reflect.DeepEqual(oldBC.Spec, newBC.Spec) || newBC.Status.LastVersion != oldBC.Status.LastVersion {
```
Comment (adambkaplan) [inline pkg/build/apiserver/registry/buildconfig/strategy.go:77]: `Status.LastVersion` gets incremented when a build is instantiated from a BuildConfig (ex `oc start-build`). Do we want to increment generation when there are no changes to `spec`? Is there value in doing this?

**Review by adambkaplan — COMMENTED** [review]
I think we need to consider whether or not Generation needs to be incremented when builds are instantiated.

**Comment by adambkaplan** [comment]
/cc @bparees 

Perhaps you know better the rules and conventions around `metadata.Generation`

File: pkg/build/apiserver/registry/buildconfig/strategy.go (hunk 1 above)
Comment (bparees) [reply pkg/build/apiserver/registry/buildconfig/strategy.go:77]: LastVersion is arguably a spec value (should have been a spec value) so treating it like one kinda makes sense.

**Comment by bparees** [comment]
> Perhaps you know better the rules and conventions around metadata.Generation

i don't have any magic answers, no.  Given how long this has been empty and no one has cared, i'm not too worried about making a wrong choice here.  I think there are good arguments for both:

1) only update it when an actual spec field is touched
or
2) also updating it when status.LastVersion is touched only because status.LastVersion really is more like a spec field, it's just not in spec.

so personally i'm fine with either.

**Comment by adambkaplan** [comment]
> also updating it when status.LastVersion is touched only because status.LastVersion really is more like a spec field, it's just not in spec.

IMO that feels more correct, since `LastVersion` impacts the behavior of build instantiation.

> Interestingly, Generation is also dealt with for DCs, imagestreams, templates in their respective controllers in OCM.

From what I can see `Generation` is used to compute the `ObservedGeneration` field in status for deploymentConfig, ImageStreams, etc. It's an upstream api convention, but there is no hard rule that says we _should_ or _must_ populate such a field [1].

[1] https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties

**Review by adambkaplan — COMMENTED** [review]
One outstanding question regarding the new `Strategy` variable for `BuildConfig`, otherwise looks good.

**Comment by adambkaplan** [comment]
Filed https://issues.redhat.com/browse/BUILD-175 to add `ObservedGeneration` to status.

**Comment by gabemontero** [comment]
> Filed https://issues.redhat.com/browse/BUILD-175 to add `ObservedGeneration` to status.

this effectively is the placeholder for the OCM side machinations (which centered around `ObservedGeneration` as @adambkaplan highlighted) I mentioned previously

i'm good to go wrt this

File: pkg/build/apiserver/registry/buildconfig/strategy.go [hunk 2]
```diff
@@ -15,6 +16,7 @@ import (
 )
 
 var (
+	Strategy = strategy{legacyscheme.Scheme, names.SimpleNameGenerator}
```
[the remark this answers is not in the document]
Comment (coreydaley-redhat) [reply pkg/build/apiserver/registry/buildconfig/strategy.go]: Removed in current push

File: pkg/build/apiserver/registry/buildconfig/strategy_test.go [hunk 3]
```diff
@@ -249,3 +249,82 @@ func TestBuildConfigLegacyStrategy(t *testing.T) {
...
+	}{
+		// 0 - empty build, prepare for create should add condition new
+		{
+			input:              &buildapi.BuildConfig{},
+			expectedGeneration: 1,
+		},
+	}
+
+	for _, test := range tests {
+		Strategy.PrepareForCreate(apirequest.NewDefaultContext(), test.input)
```
[the remark this answers is not in the document]
Comment (coreydaley-redhat) [reply pkg/build/apiserver/registry/buildconfig/strategy_test.go]: Updated to use GroupStrategy, if the tests all pass I believe this PR should be good to go.

**Comment by coreydaley-redhat** [comment]
@adambkaplan pull request has been updated per your comments about Strategy vs. GroupStrategy.  ptal

**Review by adambkaplan — APPROVED** [review]
/lgtm

**Pull request merged**